### PR TITLE
Add AutoloadSourceLocator enabling ReflectionClass and ReflectionFunction instantiation

### LINKED ISSUE #36: Implement "autoloading" source locator
In order to automatically reflect using the exact same API as the core `Reflection*` classes, we need to be able to essentially ignore the whole `SourceLocator` concept, despite it's flexibility.

So, I think it's possible to implement a new `AutoloadingSourceLocator` that takes no constructor dependencies, which, funnily enough, would probably need to rely on the internal `ReflectionClass` API, e.g.:

```php
// if it's a class we're loading...
$reflectionClass = new \ReflectionClass($identifier->getName());
$filename = $reflectionClass->getFileName();
$source = file_get_contents($filename);
// use these to make the LocatedSource...
```

Additionally, we'd need to create a "static" reflector, which would be something like:

```php
class StaticReflector
{
    public static function reflect(...)
    {
        return (new Reflector(new AutoloadingSourceLocator()))->reflect($identifier);
    }
}
```

Then we can use this reflector in the constructors or something to reflect (still a bit hazy on how to actually implement this). The downside of this is it means we do have to LOAD the classes. So I also propose making a `LoadSafeSourceLocator` interface, which doesn't have any methods, but purely indicates that it "by design" does NOT load up classes into memory. Then in the `..\Reflector\Generic` class, we can check if `$reflection implements LoadSafeSourceLocator`, and ensure that the class is not loaded beforehand, and is not loaded after reflection, to ensure the contract is obeyed.

### TL;DR

* Create new `AutoloadingSourceLocator` that uses core refleciton API to locate source code filename
* Cleanup constructors so they are empty
* Create `StaticReflector` that uses `AutoloadingSourceLocator`
* Update `Reflection*` classes to use `AutoloadingSourceLocator`
* Create `LoadSafeSourceLocator` interface, any source locator that implements this interface MUST ensure the class does not get loaded ever. These would be the `Composer`, `SingleFile` and `String` source locators, as these do not load file. `AutoloadingSourceLocator` would load the file, so MUST NOT implement the `LoadSafeSourceLocator` interface.
* This allows us to do the following (i.e. replicate almost _exactly_ the core reflection API instantiation technique):

```php
$reflectionClass = new \BetterReflection\Reflection\ReflectionClass('My\Foo\Bar');
```

### PR DESCRIPTION
**Pull request by asgrim**
Implemented the `AutoloadSourceLocator` which means we can now use "static" creation of `ReflectionClass` objects, e.g.:

```php
$reflection = ReflectionClass::createFromName('BetterReflectionTest\Fixture\ExampleClass');
```

There are caveats (we expect the autoloader to load classes from a file, i.e. using require/include), but it does actually work.

~~It also means this can actually implement the `LoadSafeSourceLocator` (to be implemented)~~

Merging this would fix #36 and would be a better solution that I initially proposed actually :grinning: 

### Todo

* [x] Rebase onto `master` when #40 is merged
* [x] Complete implementation for `ReflectionFunction`
* [x] Use static constructor not dodgy `__construct` nullable nonsense
* ~~Create a `LoadSafeSourceLocator` and write the logic to do checking~~
* ~~Implement the @todo in [Identifier.php#L45](https://github.com/Roave/BetterReflection/blob/master/src/Identifier/Identifier.php#L45)~~

### CODE REVIEW

File: src/Reflection/ReflectionClass.php [hunk 1]
```diff
@@ -18,34 +20,39 @@ class ReflectionClass implements Reflection
...
     private $filename;
 
-    private function __construct()
+    public function __construct($className = null)
     {
-        $this->declaringNamespace = null;
-        $this->methods = [];
-        $this->constants = [];
-        $this->properties = [];
+        if ($className) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Why is `null` allowed ?

File: src/Reflection/ReflectionFunction.php [hunk 2]
```diff
@@ -2,11 +2,26 @@
...
+use BetterReflection\Reflector\FunctionReflector;
+use BetterReflection\SourceLocator\AutoloadSourceLocator;
 use PhpParser\Node\Stmt\Function_ as FunctionNode;
 use PhpParser\Node\Stmt\Namespace_ as NamespaceNode;
 
 class ReflectionFunction extends ReflectionFunctionAbstract implements Reflection
 {
+    public function __construct($functionName = null)
+    {
+        if ($functionName) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionFunction.php]: Same as above

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ */
+class AutoloadSourceLocator implements SourceLocator
+{
+    private static $autoloadLocatedFile;
+
+    public function __invoke(Identifier $identifier)
+    {
+        self::$autoloadLocatedFile = null;
+
+        if ($identifier->getType()->getName() == IdentifierType::IDENTIFIER_CLASS) {
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: This logic is hard to follow: split into private methods please

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+     * Our wrapper simply records which file we tried to load and returns
+     * boolean false indicating failure
+     *
+     * @param string $path
+     * @param string $mode
+     * @param int $options
+     * @param string $opened_path
+     * @return bool
+     */
+    public function stream_open($path, $mode, $options, &$opened_path)
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: <3 PHP internals non-interfaced stream wrappers' crap.

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            'rdev' => 0,
+            'size' => 1,
+            'atime' => time(),
+            'mtime' => time(),
+            'ctime' => time(),
+            'blksize' => 4096,
+            'blocks' => 8,
+        ];
+
+        $x = array_merge(array_values($assoc), $assoc);
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: Return directly

File: src/Reflection/ReflectionClass.php (hunk 1 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Because that is how `::createFromNode` works.

doing this:

```php
$classInfo = new ReflectionClass('Foo');
```

creates autoload source locator > creates class reflector > calls reflect > locates class > loads source code > loads AST > calls `ReflectionClass::createFromNode()` > which instantiates a new `ReflectionClass` but with no parameters

File: src/Reflection/ReflectionFunction.php (hunk 2 above)
Comment (asgrim) [inline src/Reflection/ReflectionFunction.php]: Same as above

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 4 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: I do not :P

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 5 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: ah yes, left this in for testing :+1:

File: src/Reflection/ReflectionClass.php (hunk 1 above)
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: @asgrim you are changing public API to please package API? Seems wrong :-\

Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: I don't see another way around it, ideally I need two constructors.. one for internal (`::createFromNode`) use and one for public API :/

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 3 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: Refactored this, hopefully much nicer

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * Attempts to locate the specified identifier and store the result in
+     * self::$autoloadLocatedFile
+     *
+     * @param Identifier $identifier
+     */
+    private function locateIdentifier(Identifier $identifier)
+    {
+        if ($identifier->getType()->getName() == IdentifierType::IDENTIFIER_CLASS) {
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: Cache the call to `$identifier->getType()->getName()`. You should also probably make a method that matches the constant, rather than getting the constant and then comparing outside the VO

Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: Something like `$type->isClass()`

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+     * error handler temporarily.
+     *
+     * @param string $className#
+     */
+    private function locateClassByName($className)
+    {
+        if (class_exists($className, false)) {
+            $reflection = new \ReflectionClass($className);
+            self::$autoloadLocatedFile = $reflection->getFileName();
+        } else {
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: Consider returning early

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+     * have function autoloading. Therefore if it exists, we simply use the
+     * internal reflection API to find the filename. If it doesn't we can do
+     * nothing so throw an exception.
+     *
+     * @param string $functionName
+     * @throws Exception\FunctionUndefined
+     */
+    private function locateFunctionByName($functionName)
+    {
+        if (function_exists($functionName)) {
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: Reverse this conditional to throw in it, and return outside it

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 4 above)
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: Consider putting `@link` to the stream wrapper API pages here

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+/**
+ * Use PHP's built in autoloader to locate a class, without actually loading.
+ *
+ * There are some prerequisites...
+ *   - we expect the autoloader to load classes from a file (i.e. using require/include)
+ */
+class AutoloadSourceLocator implements SourceLocator
+{
+    private static $autoloadLocatedFile;
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: Docblock

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    /**
+     * Must be implemented to return some data so that calls like is_file will work
+     *
+     * @param $path
+     * @param $flags
+     * @return mixed[]
+     */
+    public function url_stat($path, $flags)
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: Consider putting `@link` to the stream wrapper API pages here

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    {
+        $assoc = [
+            'dev' => 2056,
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: WOOOOO, loads of magic constants. WTF is does it even mean? <_<

File: src/SourceLocator/AutoloadSourceLocator.php [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+     * that it cannot find the file, so we squelch the errors by overriding the
+     * error handler temporarily.
+     *
+     * @param string $className#
+     */
+    private function locateClassByName($className)
+    {
+        if (class_exists($className, false)) {
+            $reflection = new \ReflectionClass($className);
+            self::$autoloadLocatedFile = $reflection->getFileName();
```
Comment (Ocramius) [inline src/SourceLocator/AutoloadSourceLocator.php]: This could return `false`

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 11 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: It's not magic constant, it's just dummy data (see the docblock...)

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 12 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: that's okay, exception will be caught up ahead

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 4 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: yep

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 6 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: done

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 7 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: done

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 8 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: done

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 9 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: done

File: src/SourceLocator/AutoloadSourceLocator.php (hunk 10 above)
Comment (asgrim) [inline src/SourceLocator/AutoloadSourceLocator.php]: done
